infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesPods.java

```diff
@@ -199,6 +199,61 @@ public void onClose(KubernetesClientException cause) {
     }
   }
 
+  /**
+   * Subscribes to pod events and returns the resulting future, which ends when a pod event that
+   * satisfies the predicate is received.
+   *
+   * <p>Note that the resulting future is must be explicitly cancelled when its completion no longer
+   * important because of finalization allocated resources.
+   *
+   * @param name the pod name that should be watched
+   * @param predicate a function that performs pod state check
+   * @return completable future that is completed when one of the following conditions is met:
+   *     <ul>
+   *       <li>an event that satisfies predicate is received
+   *       <li>exception while getting pod resource occurred
+   *       <li>connection problem occurred
+   *     </ul>
+   *     otherwise, it must be explicitly closed
+   */
+  public CompletableFuture<Void> watchReadinessAsync(String name, Predicate<Pod> predicate) {
+    final CompletableFuture<Void> podRunningFuture = new CompletableFuture<>();
+    try {
+      final PodResource<Pod, DoneablePod> podResource =
+          clientFactory.create().pods().inNamespace(namespace).withName(name);
+      final Watch watch =
```

Comment:
I think that there should not be any problems with WebSocket connection leaks because the method documents indicate that the future must be explicitly cancelled if it is not needed and has not reached the final state.